letsencrypt/models/letsencrypt.py

```diff
@@ -78,91 +78,108 @@ def validate_domain(self, domain):
         domain = urlparse.urlparse(
             self.env['ir.config_parameter'].get_param(
                 'web.base.url', 'localhost')).netloc
         self.validate_domain(domain)
-        account_key = self.generate_account_key()
-        csr = self.generate_csr(domain)
-        acme_challenge = get_challenge_dir()
-        if not os.path.isdir(acme_challenge):
-            os.makedirs(acme_challenge)
-        if self.env.context.get('letsencrypt_dry_run'):
-            crt_text = 'I\'m a test text'
-        else:  # pragma: no cover
-            from acme_tiny import get_crt, DEFAULT_CA
-            crt_text = get_crt(
-                account_key, csr, acme_challenge, log=_logger, CA=DEFAULT_CA)
-        with open(os.path.join(get_data_dir(), '%s.crt' % domain), 'w')\
-                as crt:
-            crt.write(crt_text)
-            chain_cert = urllib2.urlopen(
-                self.env['ir.config_parameter'].get_param(
-                    'letsencrypt.chain_certificate_address',
-                    'https://letsencrypt.org/certs/'
-                    'lets-encrypt-x3-cross-signed.pem')
-            )
-            crt.write(chain_cert.read())
-            chain_cert.close()
-            _logger.info('wrote %s', crt.name)
-        reload_cmd = self.env['ir.config_parameter'].get_param(
-            'letsencrypt.reload_command', False)
-        if reload_cmd:
-            _logger.info('reloading webserver...')
-            self.call_cmdline(['sh', '-c', reload_cmd])
+        # Generate account key
+        account_key_file = self._generate_key('account_key')
+        account_key = jose.JWKRSA.load(open(account_key_file).read())
+        # Generate domain key
+        domain_key_file = self._generate_key(domain)
+        net = client.ClientNetwork(account_key)
+        if self.env.context.get('DRY_RUN'):
+            directory_url = V2_STAGING_DIRECTORY_URL
+        else:
+            directory_url = V2_DIRECTORY_URL
+        directory_json = requests.get(directory_url).json()
+        acme = client.ClientV2(directory_json, net)
+        reg = Registration(
+            key=account_key.public_key(),
+            terms_of_service_agreed=True)
+        new_reg = NewRegistration(**dict(reg))
+        acme.new_account(new_reg)
+        altnames = parameter_model.search(
+            [('key', 'like', 'letsencrypt.altname.')],
+            order='key')
+        valid_domains = [x.value for x in altnames] + [domain]
+        _logger.info(
+            'Making CSR for the following domains: ' + str(valid_domains))
+        csr = crypto_util.make_csr(
+            open(domain_key_file).read(), valid_domains)
+        authzr = acme.new_order(csr)
+        # write the file with the challenge
+        challenge_dir = _get_challenge_dir()
+        if not isdir(challenge_dir):
+            makedirs(challenge_dir)
+        # if the domain contains a wildcard, do dns validation
+        if '*' not in domain:
```

Comment:
can't we ask letsencrypt what type of verification is necessary?

Reply:
btw `domain` will never contain an asterisk, so if you really need to go this route, you'll need this check for all domain involved

Reply:
1) Yes, we can. But since the rfc allows for arbitrary challenges and letsencrypt while having some deviations from the rfc (seen here https://github.com/letsencrypt/boulder/blob/master/docs/acme-divergences.md) does seem to use the same challenges (http-01, dns-01). Http challenge is not available for domains that contain a wildcard, only dns. That's why I stick to http when I can.

2) Then why do we need wildcard support in this module?

Reply:
because you'd configure the wildcard in one of the altnames. It would make sense then to analyze this list, and collapse stuff, so that it's smart enough to just use the wildcard if we configure www.therp.nl, helpdesk.therp.nl and *.therp.nl for example